### PR TITLE
AM-450: get_case_document_metadata_api_for_doc_store_am

### PR DESCRIPTION
**Pull request by aashishaggarwal123**
JIRA link (if applicable)
https://tools.hmcts.net/jira/browse/AM-450

**Change description**
As part of securing document store work, AM team is implementing a new API on ccd-data-store for getting the document specific metadata along with users permissions on the the requested document field. This is required to secure the document download functionality.  
**Follow below LLD link for over all case document download functionality**:
https://tools.hmcts.net/confluence/pages/viewpage.action?pageId=1268384125&__ncforminfo=-duT2AqFSqDsdXviTCpcMSAhOBtdHQOZkEqeklVmt07m7mKEFA9Idv1cPubfOBvp2n8GO3X_134n0W6_c_e1LNXveh0-bpHe#CaseDocument&CCD-DataStoreDesign-DocumentDownloadProcess
**API Spec** :
https://tools.hmcts.net/confluence/pages/viewpage.action?pageId=1280607156

This PR contains:
1. New controller 'CaseDocumentController' based upon V2 pattern.
2. New service interface 'GetCaseDocumentOperation' and it's implementation 'DefaultGetCaseDocumentOperation'.
3.  New domain entities : CaseDocument, CaseDocumentMetadata & Permission
4.  New HATEOS resource for api response : CaseDocumentResource
5.  New validation service for validating document Id : DocumentIdValidationService
6.  New case document specific exceptions : CaseDocumentNotAccessibleException, CaseDocumentNotFoundException
7. new feature file for covering functional test cases based on befta framework : F-1001

Does this PR introduce a breaking change? (check one with "x")

[ ] Yes
[x] No

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+        throw new CaseNotFoundException(caseId);
+    }
+
+    public CaseDocument getCaseDocument(CaseDetails caseDetails, String documentId) {
+
+        //retrieve the case data from details
+        Map<String, JsonNode> caseData = caseDetails.getData();
+
+        //get document field name
+        String documentField = getDocumentCaseField(caseData, documentId);
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java]: if the case contains multiple fields for the same document id (technically possible and allowed in ccd backend), which field will be returned to here?

File: src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    private final GetCaseOperation getCaseOperation;
+    private final CaseTypeService caseTypeService;
+    private final UserRepository userRepository;
+    private final CaseUserRepository caseUserRepository;
+    private final DocumentIdValidationService documentIdValidationService;
+
+    public static final String DOCUMENT_CASE_FIELD_CHILD_URL = "document_url";
+    public static final String DOCUMENT_CASE_FIELD_CHILD_NAME = "document_filename";
+    public static final String DOCUMENT_CASE_FIELD_CHILD_TYPE = "Document";
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java]: DOCUMENT_CASE_FIELD_XYZ_ATTRIBUTE

File: src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    private Set<String> getUserRoles(String caseId) {
+        return Sets.union(userRepository.getUserRoles(),
+            new HashSet<>(caseUserRepository
+                .findCaseRoles(Long.valueOf(caseId), userRepository.getUserId())));
+    }
+
+    private String getDocumentCaseField(Map<String, JsonNode> caseData, String documentId) {
+        for (Map.Entry<String, JsonNode> entry : caseData.entrySet()) {
```
[the remark this answers is not in the document]
Comment (MSancaktutar) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java]: or we have a collection field of documents? or a collection of a complex type that contains a document.

File: src/main/java/uk/gov/hmcts/ccd/v2/external/resource/CaseDocumentResource.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+@Data
+@EqualsAndHashCode(callSuper = true)
+@NoArgsConstructor
+public class CaseDocumentResource extends RepresentationModel {
+
+    @JsonProperty("documentMetadata")
+    private CaseDocumentMetadata documentMetadata;
+
+    public CaseDocumentResource(@NonNull String caseId, @NonNull String documentId, CaseDocumentMetadata documentMetadata) {
```
Comment (MSancaktutar) [inline src/main/java/uk/gov/hmcts/ccd/v2/external/resource/CaseDocumentResource.java]: make sure we are returning the v1 indicator of the resource in the api responses, in the Content-Type headers.

File: build.gradle [hunk 5]
```diff
@@ -104,11 +104,11 @@ ext {
     swagger2Version = '3.0.0-SNAPSHOT'
     limits = [
             'instruction': 90,
-            'branch'     : 85,
+            'branch'     : 30,
```
[the remark this answers is not in the document]
Comment (aashishaggarwal123) [reply build.gradle]: It is updated back to 85%.

File: src/main/java/uk/gov/hmcts/ccd/v2/external/controller/CaseDocumentController.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    )
+    @ApiOperation(
+        value = "Retrieve a case document metadata by case and document Id",
+        notes = V2.EXPERIMENTAL_WARNING
+    )
+    @ApiResponses({
+        @ApiResponse(
+            code = 200,
+            message = "Success",
```
[the remark this answers is not in the document]
Comment (aashishaggarwal123) [reply src/main/java/uk/gov/hmcts/ccd/v2/external/controller/CaseDocumentController.java]: updated.

File: src/main/java/uk/gov/hmcts/ccd/v2/external/controller/CaseDocumentController.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        ),
+        @ApiResponse(
+            code = 404,
+            message = V2.Error.CASE_DOCUMENT_NOT_FOUND
+        )
+    })
+    public ResponseEntity<CaseDocumentResource> getCaseDocumentMetadata(@PathVariable("caseId") String caseId,
+                                                                        @PathVariable("documentId") String documentId) {
+
+        final CaseDocumentMetadata documentMetadata = this.getCaseDocumentOperation.getCaseDocumentMetadata(caseId,documentId);
```
[the remark this answers is not in the document]
Comment (aashishaggarwal123) [reply src/main/java/uk/gov/hmcts/ccd/v2/external/controller/CaseDocumentController.java]: Updated

File: src/main/java/uk/gov/hmcts/ccd/v2/external/domain/CaseDocument.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import lombok.NoArgsConstructor;
+
+import java.util.List;
+
+@Data
+@EqualsAndHashCode
+@NoArgsConstructor
+@AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (aashishaggarwal123) [reply src/main/java/uk/gov/hmcts/ccd/v2/external/domain/CaseDocument.java]: It seems @AllArgsConstroctor directly comes with @builder so it is not required and it doesn't like @NoArgsCOnstroctor as well, so I had to remove both.

File: src/main/java/uk/gov/hmcts/ccd/v2/external/domain/Permission.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
+package uk.gov.hmcts.ccd.v2.external.domain;
+
+/**
+ * Exposes a set of enum values used to set permissions for Access Management.
+ * Each of the values is a power of two. The reason for that is that in Access Management
+ * there might be multiple permissions that need to be assigned: ie. READ + CREATE.
+ * In order to determine which individual permissions a record has
+ * the binary 'AND' operation is done (the 'isGranted' method).
```
[the remark this answers is not in the document]
Comment (aashishaggarwal123) [reply src/main/java/uk/gov/hmcts/ccd/v2/external/domain/Permission.java]: description updated to remove reference of isGranted method. It is not required.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    public CaseDocumentMetadata getCaseDocumentMetadata(String caseId, String documentId) {
+
+        if (!documentIdValidationService.validateDocumentUUID(documentId)) {
+            throw new BadRequestException(BAD_REQUEST_EXCEPTION_DOCUMENT_INVALID);
+        }
+
+        final CaseDetails caseDetails = this.getCaseOperation.execute(caseId)
+            .orElseThrow(() -> new CaseNotFoundException(caseId));
+
+        if (!caseDetails.getReferenceAsString().isEmpty()) {
```
[the remark this answers is not in the document]
Comment (aashishaggarwal123) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java]: Thanks for noticing, I have updated.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+    private Set<String> getUserRoles(String caseId) {
+        return Sets.union(userRepository.getUserRoles(),
+            new HashSet<>(caseUserRepository
+                .findCaseRoles(Long.valueOf(caseId), userRepository.getUserId())));
+    }
+
+    private String getDocumentCaseField(Map<String, JsonNode> caseData, String documentId) {
+        for (Map.Entry<String, JsonNode> entry : caseData.entrySet()) {
+            if (entry.getValue().getNodeType().toString()
+                .equals("OBJECT") && entry.getValue().toString().contains(documentId)) {
```
[the remark this answers is not in the document]
Comment (aashishaggarwal123) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java]: Right, I have updated it to check the null check before that.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java (hunk 2 above)
Comment (aashishaggarwal123) [reply src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java]: Thanks, it makes more sense. updated

File: src/main/java/uk/gov/hmcts/ccd/v2/external/resource/CaseDocumentResource.java (hunk 4 above)
Comment (aashishaggarwal123) [reply src/main/java/uk/gov/hmcts/ccd/v2/external/resource/CaseDocumentResource.java]: New resource is added with following value :         public static final String CASE_DOCUMENT = "application/vnd.uk.gov.hmcts.ccd-data-store-api.case-document.v1+json;charset=UTF-8";

File: src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+
+        List<CaseField> documentCaseFields = new ArrayList<>();
+        List<CaseField> complexCaseFieldList = caseType.getCaseFields()
+            .stream()
+            .filter(caseField -> (DOCUMENT.equalsIgnoreCase(caseField.getFieldType().getType())) ||
+                                 (COMPLEX.equalsIgnoreCase(caseField.getFieldType().getType())) ||
+                                 (COLLECTION.equalsIgnoreCase(caseField.getFieldType().getType())))
+            .collect(Collectors.toList());
+
+        if (complexCaseFieldList.isEmpty()) {
```
Comment (nitishsingla1234) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java]: Update the name

File: src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java [hunk 13]
```diff
@@ -0,0 +1,180 @@
...
+
+    private void extractDocumentFieldsFromCaseDefinition(List<CaseField> complexCaseFieldList, List<CaseField> documentCaseFields) {
+        for (CaseField caseField : complexCaseFieldList) {
+            switch (caseField.getFieldType().getType()) {
+                case DOCUMENT:
+                    documentCaseFields.add(caseField);
+                    break;
+                case COMPLEX:
+                case COLLECTION:
+                    extractDocumentFieldsFromCaseDefinition(caseField.getFieldType().getComplexFields(), documentCaseFields);
```
Comment (nitishsingla1234) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java]: Verify the logic here.

File: src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java [hunk 14]
```diff
@@ -0,0 +1,180 @@
...
+        return CaseDocumentMetadata.builder()
+            .caseId(caseDetails.getReferenceAsString())
+            .caseTypeId(caseDetails.getCaseTypeId())
+            .jurisdictionId(caseDetails.getJurisdiction())
+            .document(getCaseDocument(caseDetails, documentId))
+            .build();
+
+    }
+
+    public CaseDocument getCaseDocument(CaseDetails caseDetails, String documentId) {
```
Comment (nitishsingla1234) [inline src/main/java/uk/gov/hmcts/ccd/domain/service/getcasedocument/GetCaseDocumentOperation.java]: private method.

**Comment by aashishaggarwal123** [comment]
This is duplicate PR.

**Pull request closed without merging**